File: 12-7-18/algorithms_challenge.py
# Challenge 1

def collatzSequence(n):

    sequence = []

    if n <= 0:
        return n
    else:
        while n != 1:
            if n % 2 == 0:
                n = n // 2
                sequence.append(n)
            else:
                n = n * 3 + 1
                sequence.append(n)
        return sequence

# print(collatzSequence(37))

# ------------------------------------------------------------------------------------

# Challenge 2

def largestPalindrome(n1, n2):

    limit1 = range(1, n1 + 1)
    limit2 = range(1, n2 + 1)
    palindromes = [i * j for i in limit1 for j in limit2 if str(i * j) == str(i * j)[::-1]]

    return max(palindromes)

# print(largestPalindrome(999, 999))

# ------------------------------------------------------------------------------------

# Challenge 3

def smallestDivisible(n):

    arr = [i for i in range(1, n + 1)]
    limit = 1
    for num in arr:
        limit = limit * num
    
    for i in range(n, limit + 1):
        check = True
        for num in arr:
            if i % num != 0:
                check = False
                break
        
        if check:
            return i
            
# print(smallestDivisible(20))

check_list = [i for i in range(1, 21)]

def find_solution(step):
    for num in range(step, 999999999, step):
        if all(num % n == 0 for n in check_list):
            return num
    return None

print(find_solution(2520))

# ------------------------------------------------------------------------------------

# Challenge 4

File: 12-7-18/algorithms_challenge.js
// Challenge 1

function collatzSequence(n) {
    
    var sequence = [];

    if (n <= 0) {
        return n;
    }
    else {
        while (n != 1) {
            if (n % 2 == 0) {
                n = n / 2;
                sequence.push(n);
            }
            else {
                n = n * 3 + 1;
                sequence.push(n);
            }
        }
        return sequence;
    }
}

// console.log(collatzSequence(17))

// ------------------------------------------------------------------------------------

// Challenge 2

function largestPalindrome (n1, n2) {

    var palindromes = [];
    
    for (var i = 1; i <= n1 + 1; i ++) {
        for (var j = 1; j <= n2 + 1; j ++) {
            var product = (i * j).toString();
            var reverse = product.split("").reverse().join("");
            if (product == reverse) {
                palindromes.push(product);
            }
        }
    }
    return Math.max.apply(null, palindromes);
}

// console.log(largestPalindrome(999, 999))

// ------------------------------------------------------------------------------------

// Challenge 3

function smallestDivisible (n) {

    var product = 1;

    for (var i = 2; i <= n; i ++) {
        product = product * i;
    }

    for (var i = n; i <= product; i++) {
        var check = true;
        for (var j = 2; j <= 20; j ++) {
            if (i % j != 0) {
                check = false;
                break;
            }
        }
        if (check) {
            return i;
        }
    }
}

// console.log(smallestDivisible(20));

// ------------------------------------------------------------------------------------

// Challenge 4

